include/wil/coroutine.h

```diff
@@ -0,0 +1,791 @@
+        auto as_promise() noexcept
+        {
+            return static_cast<Promise*>(this);
+        }
+
+        // Make it easier to access the coroutine handle.
+        auto as_handle() noexcept
+        {
+            return __WI_COROUTINE_NAMESPACE::coroutine_handle<Promise>::from_promise(*as_promise());
+        }
+
+        auto get_return_object() noexcept
+        {
+            // let the compiler construct the task / com_task from the promise.
+            return as_promise();
+        }
+
+        void destroy()
+        {
+            as_handle().destroy();
+        }
+
+        // The client lost interest in the coroutine, either because they are discarding
+        // the result without awaiting (risky!), or because they have finished awaiting.
+        // Discarding the result without awaiting is risky because any exception in the coroutine
+        // will be unobserved and result in a crash. If you want to disallow it, then
+        // raise an exception if waiting == running_ptr.
+        void abandon()
+        {
+            auto waiting = m_waiting.exchange(abandoned_ptr(), std::memory_order_acq_rel);
+            if (waiting != running_ptr()) destroy();
+        }
+
+        __WI_COROUTINE_NAMESPACE::suspend_never initial_suspend() noexcept
+        {
+            return {};
+        }
+
+        template<typename...Args>
+        void set_value(Args&&... args)
+        {
+            m_holder.set_value(wistd::forward<Args>(args)...);
+        }
+
+        void unhandled_exception() noexcept
+        {
+            m_holder.unhandled_exception();
+        }
+
+        void resume_waiting_coroutine(void* waiting) const
+        {
+            if (m_resumer)
+            {
+                m_resumer(waiting);
+            }
+            else
+            {
+                __WI_COROUTINE_NAMESPACE::coroutine_handle<>::from_address(waiting).resume();
+            }
+        }
+
+        auto final_suspend() noexcept
+        {
+            struct awaiter : __WI_COROUTINE_NAMESPACE::suspend_always
+            {
+                promise_base& self;
+                void await_suspend(__WI_COROUTINE_NAMESPACE::coroutine_handle<>) const noexcept
+                {
+                    // Need acquire so we can read from m_resumer.
+                    // Need release so that the results are published in the case that nobody
+                    // is awaiting right now, so that the eventual awaiter (possibly on another thread)
+                    // can read the results.
+                    auto waiting = self.m_waiting.exchange(completed_ptr(), std::memory_order_acq_rel);
+                    if (waiting == abandoned_ptr())
+                    {
+                        self.destroy();
+                    }
+                    else if (waiting != running_ptr())
+                    {
+                        WI_ASSERT(waiting != completed_ptr());
+                        self.resume_waiting_coroutine(waiting);

```

Comment:
I don't think task<T> currently supports symmetric transfer. Could it? That would be be a nice stack depth/perf improvement here.

Specifically, if what's waiting is a std::coroutine_handle<>, returning a std::coroutine_handle<> from await_suspend, rather than void, would mean the waiting coroutine could be executed directly without the extra stack depth of the current coroutine. See:
https://en.cppreference.com/w/cpp/coroutine/noop_coroutine
and
https://lewissbaker.github.io/2020/05/11/understanding_symmetric_transfer

Reply:
At the time this was originally written, MSVC didn't support symmetric transfer. Even after MSVC added support for symmetric transfer, the implementation was buggy. It was [finally fixed](https://devblogs.microsoft.com/cppblog/cpp20-coroutine-improvements-in-visual-studio-2019-version-16-11/) two weeks after the PR was created, but it will take time for everybody to migrate off of the older version of the compiler.

Reply:
Sounds like something that could be guarded under `#if _MSC_VER > some_version`?

Reply:
Yes, but it also creates interop issues, where you cannot reliably link together libraries built with different versions of MSVC (ODR violations).

Reply:
Yeah, an opt-in that's automatically enabled in C++20 mode (or with `/await:strict`) sounds good.